### PR TITLE
Fix issue with --target deletion dependent calculation

### PR DESCRIPTION
**Pull request by joeduffy**
The code that computed `--target` deletion dependents was not correct. It only used parent/child component relationships, but did not respect actual DAG dependencies (which are more common). As a result, it could erroneously leave hanging references to resources that no longer exist after performing a `pulumi destroy --target X` operation. This manifested in bugs like https://github.com/pulumi/pulumi/issues/6283, which is fixed by this change. The solution is to compute the (transitive!) dependency graph correctly, factoring in both parent/child, as well as explicit and implicit, dependencies. The existing logic does the correct thing once we do this. I've also added tests for this area, including regression tests that cover transitive dependency relationships, as well as ones that would cause an infinite loop given a naive implementation.

This came up during a customer call today and I was genuinely curious, and after seeing it in action, the fix seemed fairly self-evident to me, so I rolled with it. Happy to leave it to someone else to fix if my approach is not desirable.

### CODE REVIEW

**Comment by t0yv0** [comment]
This is awesome, thank you, will work through this shortly.

CC @emiliza 

Tangentially related - we have a stalled thread in https://github.com/pulumi/pulumi/pull/7852 that was working through the implications of dependency edges via Component resources. I think this is not in play here, but there's the `--target` parameter in play instead that I was not aware of. I'd like to model-check this functionality also (in addition to unit tests) just as https://github.com/pulumi/pulumi/pull/7859 was fruitful in finding bugs in 7852. 

I'd push to invest a bit of time here to get the dependency story rock-solid.

**Comment by pgavlin** [comment]
>  It only used parent/child component relationships, but did not respect actual DAG dependencies

😬 

> The solution is to compute the (transitive!) dependency graph correctly, factoring in both parent/child, as well as explicit and implicit, dependencies

Would we be able to use the preexisting [`DependencyGraph`](https://pkg.go.dev/github.com/pulumi/pulumi/pkg/v3@v3.17.0/resource/graph#DependencyGraph) code to compute this?

**Comment by joeduffy** [comment]
TIL `DependencyGraph`! @pgavlin I tried going that route, thinking I could use [`DependingOn`](https://github.com/pulumi/pulumi/blob/pkg/v3.17.0/pkg/resource/graph/dependency_graph.go#L23) -- since we are looking for dependents, not dependencies -- but that doesn't seem to include parent/child logic the same way that [`DependenciesOf` does](https://github.com/pulumi/pulumi/blob/7e2c74d58c989db04c8b877e9f4a2f7361c3b461/pkg/resource/graph/dependency_graph.go#L113-L116). I may very well be out of my league on this one; do you know if this is intentional?

**Comment by t0yv0** [comment]
Consider https://github.com/pulumi/pulumi/pull/8399

File: pkg/resource/graph/dependency_graph.go [hunk 1]
```diff
@@ -20,7 +20,8 @@ type DependencyGraph struct {
 // order with respect to the snapshot dependency graph.
 //
 // The time complexity of DependingOn is linear with respect to the number of resources.
-func (dg *DependencyGraph) DependingOn(res *resource.State, ignore map[resource.URN]bool) []*resource.State {
+func (dg *DependencyGraph) DependingOn(res *resource.State,
+	ignore map[resource.URN]bool, includeChildren bool) []*resource.State {
```
Comment (t0yv0) [inline pkg/resource/graph/dependency_graph.go:24]: I'm wondering whether this should be `includeChildren=true` always.  It's great to be conservative with the change here so we minimize blast radius, but we should follow up to figure this out what the implications here are..

Comment (joeduffy) [reply pkg/resource/graph/dependency_graph.go:24]: That was my initial instinct too, but was trying to reduce risk with the change. For what it’s worth, I tried running the unit tests with this always set to true and, somewhat terrifyingly, they passed. @pgavlin @justinvp What is the intended behavior here?

**Comment by iwahbe** [comment]
>  but that doesn't seem to include parent/child logic the same way that [`DependenciesOf` does](https://github.com/pulumi/pulumi/blob/7e2c74d58c989db04c8b877e9f4a2f7361c3b461/pkg/resource/graph/dependency_graph.go#L113-L116). I may very well be out of my league on this one; do you know if this is intentional?

[`DependenciesOf`](https://github.com/pulumi/pulumi/blob/7e2c74d58c989db04c8b877e9f4a2f7361c3b461/pkg/resource/graph/dependency_graph.go#L113-L116) only handles the immediate parent, even though the rest of the function is for transitive dependencies. I would be very curious why this does not handle transitive parents.

**Comment by t0yv0** [comment]
I'd love to merge it with https://github.com/pulumi/pulumi/pull/8399 addition which is relevant to the bug here. Transitivity seems fairly relevant to the bug here, if parent resource is included in --target then why not grandparent? But does the addendum sound too controversial?

Let me quickly check again about merging as-is.

@iwahbe thanks for you review on https://github.com/pulumi/pulumi/pull/8404 - let's finish this out and when Pat is back cross-check it so we can resolve the questions around DependenciesOf intuition.

**Comment by t0yv0** [comment]
Ok I've model checked the code in this PR against this:

```
// Model `DependingOn` as the inverse transitive closure of `dependenciesOf`.
func expectedDependingOn(universe []*resource.State, includeChildren bool) R {
	if includeChildren {

		dependingOn := inverse(transitively(universe)(
			union(hasProvider, hasDependency)))

		return func(a, b *resource.State) bool {

			if dependingOn(a, b) {
				return true
			}

			if isParent(b, a) {
				return true
			}

			for _, x := range universe {
				if dependingOn(x, b) && isParent(x, a) {
					return true
				}
			}

			return false
		}

	} else {
		return inverse(transitively(universe)(union(hasProvider, hasDependency)))
	}
}

```

In prolog notation easiest to read:

```

  dependingOnIncludingChildren(a, b) -: dependingOn(a, b).
  dependingOnIncludingChildren(a, b) -: isParent(b, a).  
  dependingOnIncludingChildren(a, b) -: dependingOn(x, b), isParent(x, a).  

```

It doesn't go into grand parents but it makes some sense, no harm in this. Approving.

**Review by t0yv0 — APPROVED** [review]

**Comment by joeduffy** [comment]
Thanks @t0yv0, fully agree with the potential broader point you’re making. The code has to compute the transitive set because the DependencyGraph wasn’t doing it (I found out because the original fix did this and when I switched to DG, it did not pass the new tests). Thanks for merging as is, and definitely a fan of doing the deeper fixing/refactoring you suggest.

**Comment by t0yv0** [comment]
Would you like a point release for this @joeduffy or should this wait for the next cycle? e.g. how urgent is the bug?

Pat should be back on Monday.

**Comment by t0yv0** [comment]
I think I see what you're saying, I missed that, since the code does a transitive set over DependingOn, the user will not observe this difference! My tests are at the wrong level so to say. Thanks for that point.

**Comment by joeduffy** [comment]
It’s ok to wait for the next release. Thank you for asking!

**Pull request merged**